src/suggestion.cpp

```diff
@@ -137,15 +137,13 @@ Xapian::Query SuggestionDataBase::parseQuery(const std::string& query)
   xquery = m_queryParser.parse_query(query, flags);
 
   if (!query.empty()) {
-    Xapian::QueryParser suggestionParser = m_queryParser;
-    suggestionParser.set_default_op(Xapian::Query::op::OP_OR);
-    suggestionParser.set_stemming_strategy(Xapian::QueryParser::STEM_NONE);
-    Xapian::Query subquery_phrase = suggestionParser.parse_query(query);
+    m_queryParser.set_stemming_strategy(Xapian::QueryParser::STEM_NONE);
```

Comment:
Same question, do we need this ?
The `m_queryParser` is already configured at init if we have found a language for the database.
So we should probably not need to reconfigure it again.

Reply:
Here, the stemming strategy STEM_NONE is correct and that set during initialization of database(STEM_SOME) is wrong. For suggestions, we need a xapian postlist for OP_PHRASE to work correctly and it is only available for unstemmed terms. So I am removing the bad initial configuration.

Reply:
But here we are "reconfiguring" the parser every time with the same value. We must configure it at init and simply use it later.
If the initial configuration is wrong it should be fixed, not removed.

Reply:
The "reconfiguring" here is intentional and it brings another bug to my notice that could surface up during caching. 

As the unit test [stemmedSearch](https://github.com/openzim/libzim/blob/master/test/suggestion_iterator.cpp#L171) indicates, we do support some stemming in suggestions via the above mentioned ["one"](https://github.com/openzim/libzim/blob/master/src/suggestion.cpp#L137) query that is causing all the fuss. So this one query, uses `STEM_SOME` to achieve this stemming following it, we then switch the stemming to `STEM_NONE` for the phrase queries.

To completely eliminate the reconfiguring of queryParser, we could have two separate queryParsers that uses the two different stemming strategies for the "one" and "Phrase" queries, but will cause a memory overhead. Is the time overhead of this reconfiguration a lot? 
I am a bit split on the best way to fix it. @mgautierfr ideas?

Reply:
In your code there is only one configuration, with always the same value.
So if you need to parse the query with two different configurations there is something wrong.
Reconfiguring the parser with the correct value just before the parsing is ok* but you need two values.

Here you parse the query one time with the default configuration, then configure it with `STEM_NONE` and do the second parsing.
If you call again `parseQuery`, the first parsing will be wrong (because made on a `STEM_NONE` configured parser).

In the previous code, we were cloning the parser and configuring the clone (so `m_queryParser` was unchanged) but as we don't clone now, we must "reset" the parser before using it again.

The solution is probably to keep the cloning into `suggestionParser` and configure the newly created `suggestionParser`.

[*] We have to be careful with multithreading. If two thread call the parser in the same time, we may have strange behavior. It is better to keep things const when possible.

Reply:
@maneeshpm Any feedback on this comment?

Reply:
@mgautierfr It seems like copying the `m_queryParser` to a `suggestionParser` and configuring the `suggestionParser` for phrase-specific operation is futile. Referring to the Xapian documentation, we can see that internally the `QueryParser` class has an intrusive pointer. So I suppose modifying either of `m_queryParsers` or the copied parser we are reconfiguring the same reference.
Even in the current git head, we were switching the `default_op` of the copied parser, but its effect was on the original `m_queryParser` hence causing all the fuss during caching.
 
I think the best way to go about it is to reconfigure the original `m_queryParser`, but this might lead to some issues in multithreading as you mentioned. Should we have two `m_queryParser`?

Reply:
Arf, you are right, the `QueryParser` is referencing the same internal value, so it is useless to copy it.
Recreating a full parser is probably a bit to complex.
The easier is probably to configure the stemming just before every (intermediate) parsing and protect the whole parsing with a mutex to protect us from race condition.

Reply:
I am implementing a mutex lock for the first time, looking forward to your review! :smile: